watcherx/changefeed.go

```diff
@@ -199,3 +202,62 @@ func WatchChangeFeed(ctx context.Context, cx *sqlx.DB, tableName string, c Event
 
 	return d, nil
 }
+
+// deduplicate sents events from `events` to the `deduplicated` channel, but
+// deduplicates events that are sent multiple times. This is necessary, because
+// the CochroachDB changefeed has a atleast-once guarantee for change events,
+// meaning that events could be sent multiple times.
+//
+// For deduplication, the last x `pastEvents` are considered.
+func deduplicate(in <-chan Event, out chan<- Event, pastEvents int) {
+	go func() {
+		defer close(out)
+		previous := newRingBuffer(pastEvents)
+
+		for {
+			e, ok := <-in
+			if !ok {
+				return
+			}
+			if previous.Contains(e) {
+				// Ignore event
+				continue
+			} else {
+				previous.Add(e)
+				out <- e
+			}
+		}
+	}()
+}
+
+type ringBuffer struct {
+	capacity int
+	seen     map[string]struct{} // map for efficient Contains().
+	keys     []string            // ring buffer so we can evict events on FIFO basis.
+	keyIdx   int                 // index of the next key to be added.
+}
+
+func newRingBuffer(capacity int) *ringBuffer {
+	return &ringBuffer{
+		capacity: capacity,
+		seen:     make(map[string]struct{}, capacity),
+		keys:     make([]string, capacity, capacity),
+	}
+}
+
+func (r *ringBuffer) Contains(el fmt.Stringer) bool {
+	_, ok := r.seen[el.String()]
+	return ok
+}
+
+func (r *ringBuffer) Add(el fmt.Stringer) {
+	// Evict the oldest key.
+	if oldestKey := r.keys[r.keyIdx%r.capacity]; oldestKey != "" {
```

Comment:
This ring buffer is not thread-safe, is there not some stdlib tooling we can use instead?

Reply:
It's not built for concurrent use, yes. The usage is just within the `deduplicate` function.

Reply:
That makes sense, in that case I think a short codedoc comment at the ringbuffer struct / instantiator or methods (Add) would be helpful for future generations

Reply:
OK, will add!